### PR TITLE
CP-46631: Add debug logs for most frequent spans.

### PR DESCRIPTION
**Pull request by GabrielBuica**
Adds debug logs of the most frequent 10 spans when the spans table is full.
If the span table is full, when adding starting a new span, forcefully finish the oldest half of the most frequent spans before adding a new one.

In case the spans table is getting full, we want to know the most frequent spans. This will help us to better understand why they are not finishing in time.

As a consequence of having a shared lock between `spans` and `finished_spans`, the previous `mark_as_finished` couldn't be called when forcibly closing the spans. My workaround was to add some optional parameters `with_lock` that would allow some of the functions to be called without locks. As it is only a workaround, I have a subsequent PR, https://github.com/xapi-project/xen-api/pull/5467, that tries to create a better abstraction for the span tables.

### CODE REVIEW

File: ocaml/libs/tracing/tracing.ml [hunk 1]
```diff
@@ -279,6 +279,56 @@ module Spans = struct
...
+
+  let most_frequent n =
+    let lst =
+      frequency_tbl
+      |> Hashtbl.to_seq
+      |> List.of_seq
+      |> List.sort (fun x y -> compare (snd x) (snd y))
+      |> List.rev
+    in
+    let rec first_n_of_lst acc n lst =
```
Comment (psafont) [inline ocaml/libs/tracing/tracing.ml:309]: This function is defined in [Xapi_stdext_std.Listext.List](https://github.com/xapi-project/stdext/blob/master/lib/xapi-stdext-std/listext.mli#L22) Please do not redefine it.

File: ocaml/libs/tracing/tracing.ml [hunk 2]
```diff
@@ -279,6 +279,56 @@ module Spans = struct
...
+        else
+          Hashtbl.replace frequency_tbl span.name (pred v)
+
+  let most_frequent n =
+    let lst =
+      frequency_tbl
+      |> Hashtbl.to_seq
+      |> List.of_seq
+      |> List.sort (fun x y -> compare (snd x) (snd y))
+      |> List.rev
```
Comment (psafont) [inline ocaml/libs/tracing/tracing.ml:307]: Sort in reverse order instead:
```
      |> List.sort (fun x y -> compare (snd y) (snd x))
```

Comment (psafont) [reply ocaml/libs/tracing/tracing.ml:307]: In general I'm not fond of sorting a list of +1000 elements to get just 10.

I would rather have a dedicated fold over the hashtable to keep the n most frequent values, probably maintained as a Map.

```
let take_most k acc ((k, freq) as elem) =
  let last_min = Map.min_elem acc in
  if freq <= last_min then
    acc
  else
    Map.remove last_min acc |> Map.add freq k
in

let most_frequent = Hashtable.fold (take_most n) Map.empty freq_tbl
```

File: ocaml/libs/tracing/tracing.ml (hunk 1 above)
[the remark this answers is not in the document]
Comment (psafont) [reply ocaml/libs/tracing/tracing.ml:309]: I agree, this is mixing responsibilities, one should be getting the k most frequent. And once that is done, print those elements, which can be done individually, then using `Printf.sprintf  "Most frequent span names are [%s]" String.concat "; " ...`

File: ocaml/libs/tracing/tracing.ml (hunk 2 above)
Comment (duobei) [reply ocaml/libs/tracing/tracing.ml:307]: Cool, that's a classic usage. I have kept it in my notes.

File: ocaml/libs/tracing/tracing.ml [hunk 3]
```diff
@@ -281,6 +281,99 @@ module Spans = struct
...
+      match last_min with
+      | None ->
+          SpanMap.add k v acc
+      | Some last_min ->
+          if SpanMap.cardinal acc < n then
+            SpanMap.add k v acc
+          else if v <= snd last_min then
+            acc
+          else
+            SpanMap.remove (fst last_min) acc |> SpanMap.add k v
```
Comment (lindig) [inline ocaml/libs/tracing/tracing.ml:327]: This could use `Map.update`.

**Review by lindig — APPROVED** [review]

File: ocaml/libs/tracing/tracing.ml (hunk 3 above)
Comment (GabrielBuica) [reply ocaml/libs/tracing/tracing.ml]: At first I thought I could use it as well but I am replacing the binding with a new one with a different key.

File: ocaml/tests/test_observer.ml [hunk 4]
```diff
@@ -514,6 +596,7 @@ let test =
     , `Quick
     , test_observer_create_and_destroy
     )
+  ; ("test_most_frequent_spans", `Quick, test_most_frequent_spans)
```
Comment (GabrielBuica) [inline ocaml/tests/test_observer.ml:599]: If I put this as a later test, it fails. The issue being the tables are not empty at the beginning of the test.

Comment (psafont) [reply ocaml/tests/test_observer.ml:599]: That means they are not unit tests currently. Tests should be independent of each other whenever possible

**Comment by GabrielBuica** [comment]
After some talks with @edwintorok, this needs further refactoring. Closing it for now.

**Pull request closed without merging**